agent/lib/kontena/observer.rb

```diff
@@ -43,33 +68,92 @@ def values
         @observables.map{|observable| @values[observable] }
       end
 
-      # Yield to block with values
+      # Still accepting updates?
+      # @return [Boolean]
+      def active?
+        fail NotImplementedError
+      end
+
+      # Run block or resume task
       def call
-        @block.call(*values)
+        fail NotImplementedError
+      end
+
+      class Async < Observe
+        # @param actor [Celluloid::Proxy::Cell] Observing actor
+        # @param block [Block] Observing block
+        def initialize(cls, observables, &block)
+          super(cls, observables)
+          @block = block
+        end
+
+        # Persistent, expected to be updated multiple times
+        # @return [Boolean]
+        def active?
+          true
+        end
+
+        def call
+          Thread.current[:celluloid_actor].task(:observe) do
+            @block.call(*values)
+          end
+        end
+      end
+
+      class Sync < Observe
+        # @param task [Celluloid::Task] Observing task
+        def initialize(cls, observables, task)
+          super(cls, observables)
+          @task = task
+          @done = false
+        end
+
+        # Single-use
+        # @return [Boolean]
+        def active?
+          !@done
+        end
+        def done!
+          @done = true
+          @task = nil
+        end
+
+        # Run block or resume task
+        def call
+          @task.resume(self)
+        end
       end
     end
 
-    # Observe has been updated, call if ready.
-    # Called from observe as an async task.
+    # Register celluloid actor handler for Kontena::Observable::Message.
     #
-    # @param observe [Observer::Observe]
-    def observed(observe)
-      observe.call if observe.ready?
+    # Updates the Observe, and calls if ready.
+    def register_observer_handler
+      @observer_handler ||= Thread.current[:celluloid_actor].handle(Kontena::Observable::Message) do |message|
```

Comment:
This needs to get memoized in some per-actor storage, not a per-instance variable. If multiple classes called by the same actor use `Kontena::Observer`, they would register multiple handlers... alternatively, the `handle` pattern would need to be something thats per-observer?